pkg/controller/pkg/manager/reconciler.go

```diff
@@ -0,0 +1,402 @@
+	nr := func() v1alpha1.PackageRevision { return &v1alpha1.ProviderRevision{} }
+	nrl := func() v1alpha1.PackageRevisionList { return &v1alpha1.ProviderRevisionList{} }
+
+	r := NewReconciler(mgr,
+		WithInstallNamespace(namespace),
+		WithNewPackageFn(np),
+		WithNewPackageRevisionFn(nr),
+		WithNewPackageRevisionListFn(nrl),
+		WithLogger(l.WithValues("controller", name)),
+		WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor("provider"))),
+	)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.Provider{}).
+		Owns(&v1alpha1.ProviderRevision{}).
+		Complete(r)
+}
+
+// SetupConfiguration adds a controller that reconciles Configurations.
+func SetupConfiguration(mgr ctrl.Manager, l logging.Logger, namespace string) error {
+	name := "packages/" + strings.ToLower(v1alpha1.ConfigurationGroupKind)
+	np := func() v1alpha1.Package { return &v1alpha1.Configuration{} }
+	nr := func() v1alpha1.PackageRevision { return &v1alpha1.ConfigurationRevision{} }
+	nrl := func() v1alpha1.PackageRevisionList { return &v1alpha1.ConfigurationRevisionList{} }
+
+	r := NewReconciler(mgr,
+		WithInstallNamespace(namespace),
+		WithNewPackageFn(np),
+		WithNewPackageRevisionFn(nr),
+		WithNewPackageRevisionListFn(nrl),
+		WithLogger(l.WithValues("controller", name)),
+		WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor("configuration"))),
+	)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.Configuration{}).
+		Owns(&v1alpha1.ConfigurationRevision{}).
+		Complete(r)
+}
+
+// NewReconciler creates a new package reconciler.
+func NewReconciler(mgr ctrl.Manager, opts ...ReconcilerOption) *Reconciler {
+	r := &Reconciler{
+		client: resource.ClientApplicator{
+			Client:     mgr.GetClient(),
+			Applicator: resource.NewAPIPatchingApplicator(mgr.GetClient()),
+		},
+		log:    logging.NewNopLogger(),
+		record: event.NewNopRecorder(),
+	}
+
+	for _, f := range opts {
+		f(r)
+	}
+
+	return r
+}
+
+// Reconcile package.
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) { // nolint:gocyclo
+	log := r.log.WithValues("request", req)
+	log.Debug("Reconciling")
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	p := r.newPackage()
+	if err := r.client.Get(ctx, req.NamespacedName, p); err != nil {
+		// There's no need to requeue if we no longer exist. Otherwise we'll be
+		// requeued implicitly because we return an error.
+		log.Debug(errGetPackage, "error", err)
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGetPackage)
+	}
+
+	// Get existing package revisions.
+	prs := r.newPackageRevisionList()
+	if err := r.client.List(ctx, prs, client.MatchingLabels(map[string]string{parentLabel: p.GetName()})); err != nil {
+		log.Debug(errListRevisions, "error", err)
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errListRevisions)
```

Comment:
I believe `kmeta.IsNoMatchError` is returned if the API types is not able to be recognized by the client. The doc on `List` says:
```
// List retrieves list of objects for a given namespace and list options. On a
// successful call, Items field in the list will be populated with the
// result returned from the server.
``` 
Which sound to me like it should return an error if none are found. However, I actually don't know how what I have right here would be working if that was the case because we would never get past this point because we would start with no package revisions, and then we would never create any :thinking: